url47e6aca666_pocorall_automaton_tgz-pJ8-net_pocorall_automaton_DatatypesJ8/src/net/pocorall/automaton/ShuffleOperations.java

```diff
@@ -0,0 +1,335 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the author may not be used to endorse or promote products
+ *    derived from this software without specific prior written permission.
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package net.pocorall.automaton;
+
+import java.util.*;
+
+/**
+ * Automata operations involving shuffling.
+ */
+final public class ShuffleOperations {
+
+	private ShuffleOperations() {
+	}
+
+	/**
+	 * Returns an automaton that accepts the shuffle (interleaving) of
+	 * the languages of the given automata.
+	 * As a side-effect, both automata are determinized, if not already deterministic.
+	 * Never modifies the input automata languages.
+	 * <p/>
+	 * Complexity: quadratic in number of states (if already deterministic).
+	 * <p/>
+	 * <dl><dt><b>Author:</b></dt><dd>Torben Ruby
+	 * &lt;<a href="mailto:ruby@daimi.au.dk">ruby@daimi.au.dk</a>&gt;</dd></dl>
+	 */
+	public static DefaultAutomaton shuffle(DefaultAutomaton a1, DefaultAutomaton a2) {
+		BasicOperations.determinize(a1);
+		BasicOperations.determinize(a2);
+
+		Transition[][] transitions1 = LinkedAutomaton.getSortedTransitions(a1.getStates());
+		Transition[][] transitions2 = LinkedAutomaton.getSortedTransitions(a2.getStates());
+		DefaultAutomaton c = new DefaultAutomaton();
+		LinkedList<StatePair> worklist = new LinkedList<StatePair>();
+		HashMap<StatePair, StatePair> newstates = new HashMap<StatePair, StatePair>();
+		State s = new State();
+		c.initial = s;
+		StatePair p = new StatePair(s, a1.initial, a2.initial);
+		worklist.add(p);
+		newstates.put(p, p);
+		while (worklist.size() > 0) {
+			p = worklist.removeFirst();
+			p.s.accept = p.s1.accept != null && p.s2.accept != null ? Object.class : null;
+			Transition[] t1 = transitions1[p.s1.number];
+			for (int n1 = 0; n1 < t1.length; n1++) {
+				StatePair q = new StatePair(t1[n1].to, p.s2);
+				StatePair r = newstates.get(q);
+				if (r == null) {
+					q.s = new State();
+					worklist.add(q);
+					newstates.put(q, q);
+					r = q;
+				}
+				p.s.transitions.add(new Transition(t1[n1].min, t1[n1].max, r.s));
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 71:81, 83:93.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.